File: src/Indigo/Core/SelectionRange.hs
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NoFieldSelectors #-}

{-# OPTIONS_GHC -Wno-name-shadowing #-}

module Indigo.Core.SelectionRange
  ( SelectionRange

    -- * Create
  , fromPosition
  , fromPositions

    -- * Query
  , start
  , end
  , isReduced
  , isOverlapping

    -- * Modify
  , merge

    -- * Consume
  , toPositions

    -- * Internal
  , isValid
  , isStartBeforeEnd
  )
where

import Data.Default.Class (Default (..))
import Data.IntervalMap.Generic.Interval (Interval (..))
import Indigo.Core.Position (Position (..))
import Prelude hiding (flip)

data SelectionRange = SelectionRange
  { start :: !Position
  , end :: !Position
  }
  deriving stock (Show, Eq, Ord)

instance Default SelectionRange where
  def :: SelectionRange
  def = fromPosition def

instance Interval SelectionRange Position where
  lowerBound :: SelectionRange -> Position
  lowerBound selectionRange = selectionRange.start

  upperBound :: SelectionRange -> Position
  upperBound selectionRange = selectionRange.end

fromPosition :: Position -> SelectionRange
fromPosition position =
  SelectionRange
    { start = position
    , end = position
    }

fromPositions :: Position -> Position -> SelectionRange
fromPositions position1 position2 =
  SelectionRange
    { start = min position1 position2
    , end = max position1 position2
    }

start :: SelectionRange -> Position
start selectionRange = selectionRange.start

end :: SelectionRange -> Position
end selectionRange = selectionRange.end

isReduced :: SelectionRange -> Bool
isReduced selectionRange = selectionRange.start == selectionRange.end

isOverlapping :: SelectionRange -> SelectionRange -> Bool
isOverlapping = overlaps

merge :: SelectionRange -> SelectionRange -> SelectionRange
merge selectionRange1 selectionRange2 =
  SelectionRange
    { start = min selectionRange1.start selectionRange2.start
    , end = max selectionRange1.end selectionRange2.end
    }

toPositions :: SelectionRange -> (Position, Position)
toPositions selectionRange =
  ( selectionRange.start
  , selectionRange.end
  )

isValid :: SelectionRange -> Bool
isValid selectionRange = isStartBeforeEnd selectionRange

isStartBeforeEnd :: SelectionRange -> Bool
isStartBeforeEnd selectionRange = selectionRange.start <= selectionRange.end

File: src/selection_range.rs
use crate::{position::Position, range::Range};
use std::cmp::{max, min};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct SelectionRange {
    start: Position,
    end: Position,
}

impl SelectionRange {
    pub fn start(&self) -> &Position {
        &self.start
    }

    pub fn end(&self) -> &Position {
        &self.end
    }

    pub fn is_reduced(&self) -> bool {
        self.start == self.end
    }

    pub fn is_overlapping(&self, other: &Self) -> bool {
        self.start <= other.end && other.start <= self.end
    }

    pub fn merge(&self, other: &Self) -> Self {
        Self {
            start: min(self.start.clone(), other.start.clone()),
            end: max(self.end.clone(), other.end.clone()),
        }
    }

    pub fn is_valid(&self) -> bool {
        self.start <= self.end
    }
}

impl From<Position> for SelectionRange {
    fn from(position: Position) -> Self {
        Self {
            start: position.clone(),
            end: position.clone(),
        }
    }
}

impl From<(Position, Position)> for SelectionRange {
    fn from((position1, position2): (Position, Position)) -> Self {
        Self {
            start: min(position1.clone(), position2.clone()),
            end: max(position1.clone(), position2.clone()),
        }
    }
}

impl From<Range> for SelectionRange {
    fn from(range: Range) -> Self {
        Self::from((range.anchor().clone(), range.cursor().clone()))
    }
}

impl From<SelectionRange> for (Position, Position) {
    fn from(range: SelectionRange) -> Self {
        (range.start, range.end)
    }
}
